Dynamic Annotations

MichaelBarnett opened the pull request:
## Background
This update gives Commec a new suite of functionality - the ability to parse Regulation Annotations with a regional context. This allows the Taxonomy search to FLAG based on TaxID, GenBank, or Uniprot accession on a per Regulation List basis. Lists can be excluded based on `Regional Context` and their input completely ignored, or used democratically to WARN, or FLAG hits from Taxonomy Searches.

Furthermore, this functionality can be explored and interrogated using the add `commec list` cli commands.

## Changes
* Commec `Regulation` module added, a standalone module that handles list ingestion, and API for interrogating processed lists.
* `commec list` command line interface allows quick interrogation of any accession when directed to the annotated regulation list database location. As well as providing regional context inputs, and .csv summary outputs.

Example:
```
(commec-dev) root@Michael-Laptop:~/repo/working/common-mechanism# commec list -d ../../commec-databases/commec-dbs/regulated_lists/ -o ../list.csv -l -a 11084
 The Common Mechanism : List
────────┐
ERROR   │ 51 imported regulated annotations were bad entries with no TaxID, Genbank, or Uniprot Accession:
        │                                                         name  category list_acronym
        │                                Paracoccidioides brasiliensis Eukaryota        COSHH
        │                     Sporadic Creutzfeldt-Jakob disease agent  Proteins        COSHH
        │                                Sporadic fatal insomnia agent  Proteins        COSHH
        │                Variably protease-resistant prionopathy agent  Proteins        COSHH

 ... manually truncated for brevity ...

        │                                           Neosaxitoxin (NEO)    Toxins       UKSECL
        │ Novel coronaviridae (eg bat coronaviruses WIVI or SHC011)...   Viruses        COSHH
        │  Run in --verbose mode for raw row input details.
WARNING │ The following imported regulated annotations were duplicates with differing metadata:
        │ accession                                           name  category list_acronym
        │     54388                         Salmonella paratyphi A  Bacteria        ATCSA
        │      5667                        Leishmania brasiliensis Eukaryota        COSHH
        │      5666                              Naegleria fowleri Eukaryota        COSHH
        │    565995                              Reston ebolavirus   Viruses        COSHH
        │     11084 Central European tick-borne encephalitis virus   Viruses        COSHH
        │     11084                                Hanzalova virus   Viruses        COSHH
        │     11084                                     Hypr virus   Viruses        COSHH
        │     11084         Siberian tick-borne encephalitis virus   Viruses        COSHH
        │     11276                   Vesicular stomatitis Alagoas   Viruses        COSHH
        │     11276                   Vesicular stomatitis Indiana   Viruses        COSHH
        │     11276                Vesicular stomatitis New Jersey   Viruses        COSHH
INFO    │  *----------* REGULATION LISTS *----------*
INFO    │ The following Regulation Lists have been identified:
        │ [SAPO] Specified Animal Pathogen Order - United Kingdom
        │ (https://www.hse.gov.uk/pubns/books/hsg280.htm)
        │ Regulated Taxid Entries: 34, Status : Compliance
        │ [ATCSA] Schedule 5 of the Anti-terrorism, Crime and Security Act 2001 - United Kingdom
        │ (https://www.legislation.gov.uk/ukpga/2001/24/2002-05-31/data.pdf)
        │ Regulated Taxid Entries: 66, Status : Compliance
        │ [COSHH] Control of Substances Hazardous to Health Regulations 2002 - United Kingdom
        │ (https://www.hse.gov.uk/pubns/misc208.htm)
        │ Regulated Taxid Entries: 172, Status : Compliance
        │ [UKSECL] UK Strategic Export Control List - United Kingdom
        │ (https://assets.publishing.service.gov.uk/media/660d281067958c001f365abe/uk-strategic-export-control-list.pdf)
        │ Regulated Taxid Entries: 155, Status : Compliance
        │ [CLBA] Combined list of biological agents update August 2024 - The Netherlands
        │ (https://www.bureaubiosecurity.nl/documenten/combined-list-of-biological-agents-update-august-2024)
        │ Regulated Taxid Entries: 0, Status : Compliance
        │ [DUIECLPRC] Dual-Use Items Export Control List of the People’s Republic of China - People's Republic of China
        │ (https://www.china-briefing.com/news/china-issues-new-export-control-regulations/)
        │ Regulated Taxid Entries: 0, Status : Compliance
        │ [SCOMET] India Export Controls (SCOMET List) - India
        │ (https://www.mea.gov.in/Portal/Images/SCOMET-List-2021.pdf)
        │ Regulated Taxid Entries: 0, Status : Compliance
        │     [Total number of Taxid Relationships:17058]
INFO    │  *----------* REGULATED TAXIDS *----------*
INFO    │ Regulation Annotations for supplied taxids (#1):
INFO    │    > Taxid 11084: Regulated by the following lists:
        │    > Viruses Tick-borne encephalitis virus (Russian Spring-Summer encephalitis virus) regulated by Schedule 5
        │ of the Anti-terrorism, Crime and Security Act 2001 [ATCSA]
        │    > Viruses Absettarov virus regulated by Control of Substances Hazardous to Health Regulations 2002 [COSHH]
INFO    │ Writing output list summary to "../list.csv.csv" ...
────────┘
```

### New features
* Adds regulation annotation to logging during taxonomic screening steps.
* Adds user ability to interrogate lists directly through commec cli and looking at outputs. Or through generating a .csv containing the condensed list information.

### Breaking changes 
* Additional yaml configuration inputs for regional context, and for regulation annotation database location. Defaults will work if user has latest updates from commec-databases.
* Updates the JSON output version, with additional regulation list information in the commec info, and taxonomic hit information levels.

Commit:
initial planning stage implementation for regulation annotation lists.

Commit:
Reorganised where clear() lives, and initial implementation of how taxid retrieval recursive should work.

Commit:
Round of updates to the design, reinclusion of the child mapping to regulated taxid. Ready for implementing cli.

Commit:
Simply added cli entry point for commec list.

Commit:
Updated cli to add correct args, as well as initiate logging for debugging.

Commit:
Various bug fixes to import scripts.

Commit:
Added some basic summary info as debug for import of lists.

Commit:
Added simply child taxid regulation retrieval, as well as some light bug fixes regarding that process - change the dtype of taxid on import, should always be dealing with ints.

Commit:
Updated regions parsing, added regions to input yaml.

Commit:
Refactored the list print into its own function, addressed partial futureWarning with the child taxa LUT file panda dataframe query by assigning types to all columns, columns with NaNs now affecting the same warning for the regulation taxids dataframe.

Commit:
Fixed issues with FutureWarnings by mis-using pandas concat, removed hard coded column names from multiple places, integrated --regions as a way to pass regional context for list import. Working as intended.

Commit:
Tidied region.py, added specific alpha-3 checking support before reverting to search, included 3 limit in the fuzzy search allowance, which in particular allows United States as a search term.

Commit:
Tidied initialisation.py

Commit:
Treat regulation as module, updated safety regarding list acroynm overlap, include tests in overlap checks. General code tidying.

Commit:
Fixed int str taxid bug for taxid retreival, added cli options for taxid info printing. Tidied outputs.

Commit:
Updated yaml input for regions to be under regulation_lists, ensured all is lower case

Commit:
Incorperated regulation detection based off presence of taxid in the annotations list.

Commit:
Fixed small logging bug where keyword all was being treated as a search term for country search.

Commit:
Updated pandas database use for indexing - to better optimise taxid retrieval.

Commit:
Initial implementation of a more robust hashing retrieval to support genbank, uniprot, and taxid accesions. Is easily extendable for other accesion types in the future.

Commit:
refactored cli for regulations to its own file, minor spelling changes, updates the system for better use with mutliple accession formats.

Commit:
Updated to included csv cli, support for genbank and uniprot accessions extended, minor bug fixes and logging support during load, additional tests for duplicate list entrants.

Commit:
More controlled logging on bad entrants to the regulation annotation import.

Commit:
Added tests for accession identity regex testing. Checks for incorrect int inputs and assumes taxid.

Commit:
Fixed bad pandas declaration for bad entries, added json output of the list information, fixed pycountry dependancy not added to environment.yaml

MichaelBarnett commented:
TODO:
* Merge with Extra Taxonomic Info branch and synchronize new testing environment.
* Add Taxa list acronym info to each taxonomy hit output.

Commit:
Updated tests to pass under current conditions, main issues was with json version, and addition of list info to the functional json. Updated the regulation tests to use a valid ListMode instead of int.

Commit:
Added regions to the cli for commec screen, to override regional context from yaml.

alexanian commented on commec/config/result.py:
```diff
@@ -680,6 +681,8 @@ class ScreenRunInfo:
     time_taken: str = ""
     date_run: str = ""
     search_tool_info: SearchToolInfo = field(default_factory=SearchToolInfo)
+    regulation_list_info : list[RegulationList] = field(default_factory=list)
```
I have convinced myself over the course of this conversation that we should call these "Control Lists" both here and in that repo, since we want to be able to include things that are not just national regulations, a company could have a cute lil personal control list

alexanian commented on commec/regulation/containers.py:
```diff
@@ -0,0 +1,383 @@
+    """
+    TAXID = "TaxonomyID"
+    GENBANK = "GenbankProtein"
+    UNIPROT = "UniprotID"
```
Let's remove these since we don't actually support them yet, and I don't want anyone (including me in 3 months) to be misled

alexanian commented on commec/regulation/containers.py:
```diff
@@ -0,0 +1,383 @@
+    """
+    Supported hashable accesion formats.
+    """
+    TAXID = "TaxonomyID"
```
Do you know the most standard names for these? We're definitely only talking about the NCBI taxonomy database (that is what we ask people to download) - looking at [this page](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=186536&lvl=3&lin=f&keep=1&srchmode=5&unlock) maybe the term we use in code is "taxid" and in text is "NCBI Taxonomy ID"?

alexanian commented on commec/regulation/containers.py:
```diff
@@ -0,0 +1,383 @@
+@dataclass
+class Region:
+    """
+    Name and Acronym for regional information as either 
```
Can we specifically validate / use mostly ISO country codes? (even though we allow some extras; maybe worth adding in this docstring / restricting format?)

alexanian commented on commec/regulation/containers.py:
```diff
@@ -0,0 +1,383 @@
+    def __repr__(self):
+        return self.name + " ["+self.acronym+"]"
+
+class ListMode(StrEnum):
```
```suggestion
class ControlMode(StrEnum):
    """
    Describes whether to treat Taxonomy IDs on a control list as regulated.
    Typically depends on the regional context.
    
    * REGULATED - All taxids from this list are regulated.
    * CONSENSUS - Taxids are regulated if they appear in more than 1 of these lists.
    * IGNORE - Ignore this list entirely.
    """
    REGULATED = "Regulated"
    CONSENSUS = "Used to determine default regulation"
    IGNORE = "Ignored"
```

Agree that we do want to support an "Ignore" option for lists from outside a region, e.g. could have a `--region US` vs. `--region US --only` CLI options

alexanian commented on commec/regulation/containers.py:
```diff
@@ -0,0 +1,383 @@
+    target : str = ""
+    hazard_group : str = ""
+    uniprot : str = ""
+    genbank_protein : str = ""
```
I think `commec` can ignore some of these columns (e.g. notes, hazard_group, uniprot, genbank) for the moment, and would suggest we remove from the class

alexanian commented on commec/regulation/containers.py:
```diff
@@ -0,0 +1,383 @@
+# Copyright (c) 2021-2024 International Biosecurity and Biosafety Initiative for Science
```
I vaguely favour splitting this into a `dataclasses.py` and a `globals.py` if we can w/o circularity

alexanian commented on commec/regulation/initialisation.py:
```diff
@@ -0,0 +1,268 @@
+
+def update_regional_context(
+    regions_of_interest : set[str] = None,
+    alternative_mode = ListMode.CONDITIONAL_NUM
```
```suggestion
    default_outside_region = ListMode.CONDITIONAL_NUM
```

alexanian commented on commec/regulation/region.py:
```diff
@@ -0,0 +1,148 @@
+        return
+
+    region_array = []
+    global REGION_DATA_LUT
```
having an itchy feeling of "why globals... maybe should have classes" but maybe I am too object-oriented

alexanian commented on commec/tools/blast_tools.py:
```diff
@@ -14,6 +14,8 @@
+from commec.regulation.containers import AccessionFormat
 
 TAXID_SYNTHETIC_CONSTRUCTS = 32630
 TAXID_VECTORS = 29278
```
should we move these to `constants` while we're here?

alexanian commented on commec/screen.py:
```diff
@@ -231,6 +236,14 @@ def add_args(parser: argparse.ArgumentParser) -> argparse.ArgumentParser:
+        dest="regions",
+        nargs="+",
+        default=[],
+        help="A list of countries or regions to add context to list compliance i.e. NZ US CH",
```
could include more about valid region_group values for this (e.g. Australia Group) somewhere...

alexanian reviewed (commented):
I know we're still drafting and i apparently have lots of thoughts but support merging in a version of this that handles basically all the list-importing logic, then leaving the commec business logic changes for another PR

Commit:
Updated list acronym interpretation to include lists of list acroynms formatted specifically within quotation marks, as well as improved the debug for when a list acronym is not recognised, to better identify where it came from

Commit:
Updates for additional initial debugging, as well as protections for multiple list acronyms in a column.

Commit:
Minor updates to doc string of imports, removed print debug statement from screen.

Commit:
Changes warning to debug status for messages regarding shared meta data for a control list.

Commit:
Initial rename from regulation to control_list

Commit:
Fixes some issues with the control list rename

Commit:
Refactor and renaming of Regulation into Control Lists, proper module handling, Accession use streamlined, currently only supports taxids, genbank and uniprot removed. Several readiblity and code enhancements, final step before implementing the additional changes to the JSON outputs.

Commit:
Initial implementation of taxa into JSON.

Commit:
Updated functional test to use a default control list annotation.

Commit:
Altered how the regulation is captured, included necessary information outputs for a rationale text generation for each control outcome, i.e. from child etc. Added the requirement for the ignored_annotations.csv file, and started support for that, however currently has no real implementation in check reg path.

Commit:
Added functio doc string.

Commit:
Updated the regions output in the json to comprehensively list the regions involved in region definitions for better human parsing. Updates to how non regulated maintain their annotation status if they were under conditional compliance, as well as hit description.

Commit:
Corrected tests for new contro list layout, as well as fixed a minor taxid number issue that was confounding list annotation outputs.

Commit:
Slightly improved External controlled messages.

Commit:
Tidied how the screen prints out list information when initialised, tidied some type issues with list import, fixed some broken cli issues with commec list when providing an accession based on the new outputs needed by screen with get_regulation()

Commit:
Updated the conditional number constant to be directly representative of the number by using the >= logic operator, for better human readibility.

Commit:
initial implementation of the removal of using pytaxonkit, which involves the removal of using species,kingdom etc in all outputs, and a focus on control list inputs instead. Regenerated json functional test after successful checking that the output is correct, albeit with missing taxonomy extra info.

Commit:
tidied the outputs for control list logging, still need to better control and update empty category and or species strings being output to logs with no information.

Commit:
Initial changes to how the domain text is logged, however signifant logging reworks are required to switch into control list mode

Commit:
Fixed isses with carrying over category into domain text for functional test.

Commit:
Rebased onto most recent develop 1.0.2, updated version codes to 1.1.0 for upcoming release.

Commit:
Added new use column import for ControlLists, uses this for future logic, as well as to generate the output string for the control list, updates ControlLists to have a single region, as multiple regions are handled by custom regions, like EU, AG. Which also simplifies several region grabbing code snippets. Better handles error generation on ControlList parsing of region and use data.

Commit:
movement away from regulation and towards controlled verbiage.

Commit:
Updated the commec list outputs to use the new descriptions and str acronym conversions for a control list.

Commit:
updated the uses for control list enum with restricted confirmed uses

alexanian commented on commec/config/result.py:
```diff
@@ -220,9 +221,9 @@ class TaxonomyAnnotation:
-    superkingdom: str = ""
+    #species: str = ""
+    #genus : str = ""
+    #superkingdom: str = ""
```
```suggestion
```

best to delete? but we should add this info to our lists in a future version?

alexanian commented on commec/control_list/cli.py:
```diff
@@ -0,0 +1,141 @@
+        default=False,
+        action="store_true"
+    )
+    parser_obj.add_argument(
+        "-l",
+        "--list",
+        dest="showlists",
+        action="store_true",
+        help="Print a summary of all imported Control Lists",
+    )
```
I think this is just what the default should be if no flags are provided (vs. the current)

```
(/mnt/data/conda-envs/commec-dev) [ec2-user@ip-172-31-93-20 cm-dbs]$ commec list -d ~/cm-dbs/
 The Common Mechanism : List
────────┐
ERROR   │ commec list requires --lists/-l or --accessions/-a as input.
```

Commit:
Fix bug for using regions.

alexanian reviewed (commented):
A few tiny comments, working on a full review but wanted to submit these instead of leaving them until Monday

Commit:
Updated default names expected for control lists to control, not regulated.

Commit:
Updated column names for child taxid lists to contain controlled_taxid column name.

Commit:
Added lineage import.

Commit:
minor fix for control list import test.

Commit:
In progress tidying of taxonomy imports.

manu-script commented on commec/config/result.py:
```diff
@@ -34,14 +34,15 @@
 
 # Seperate versioning for the output JSON.
-JSON_COMMEC_FORMAT_VERSION = "0.3"
+JSON_COMMEC_FORMAT_VERSION = "0.4"
```
Just trying to understand the rationale for versioning of JSON output - is it important for downstream parsing using `commec flag`? In other words, is `commec flag` not backwards compatible with older JSON outputs?

manu-script commented on commec/control_list/control_list.py:
```diff
@@ -0,0 +1,187 @@
+def get_regulation(accession : str, accession_fmt : data.AccessionFormat) -> list[tuple[RegulationList, TaxidRegulation]]:
+    """
+    Check the given Accession against all imported regulated lists.
+    The input Accession can be a TaxID, GenBank protein, or Uniprot ID.
```
Is there a strong use case for including support for sequence accessions from multiple databases in the `control list` functionality?

alexanian replied:
Yeah, @manu-script would love your take on this; at present, we're only supporting lists of taxa, but we think at some point we should expand this to load up annotations for the biorisk... but as @MichaelBarnett and I are talking about it right now, we're not sure that makes sense.

I continue to think we should rephrase this as "taxid" rather than "accession" but I am willing to be outvoted

manu-script replied:
I'm also in favor of limiting the scope of `commec list` to taxids for this PR to keep it simple.

IMO, there are usually multiple GenBank/UniProt accessions for a given controlled toxin, but only the representative accession is annotated in the control lists for practical reasons (based on my limited understanding of how the control lists were annotated with accessions). It would not be ideal to return "accession not found in the control list" if a user queries a different but valid accession for a controlled toxin.

manu-script commented on commec/screen-default-config.yaml:
```diff
@@ -23,6 +23,9 @@ databases:
   taxonomy:
       path: "{default}taxonomy/"
+  regulated_lists:
+      path: "{default}regulated_lists"
```
Should we use `control_lists` instead of `regulated_lists`? Are they the same or different by definition?
```suggestion
  control_lists:
      path: "{default}control_lists"
```

MichaelBarnett replied:
Ah yes, there's a rename in progress for all instances of regulated lists, to be converted to control lists. due to some file name expectations, this is one of the last things changed.

manu-script commented on commec/control_list/cli.py:
```diff
@@ -0,0 +1,141 @@
+def add_args(parser_obj: argparse.ArgumentParser) -> argparse.ArgumentParser:
+    """
+    Add Control List module arguments to an ArgumentParser object.
+    """
```
We could also add `-y, --config CONFIG_YAML` option to read the `commec-dbs` base path if we are going to distribute the `control_lists` folder in the core database from the next release.

MichaelBarnett replied:
Yes I like this idea - originally I wanted to pass the config yaml, but then figured it was overkill and just point to the folder (what if someone was comparing outputs from two different control lists or versions, for example, and pointed it to different directories) But it should really support interpolating from a config file too for some users ease of use.

Commit:
Merge branch 'develop' into dynamic-annotations

Commit:
Updated the generated examples which had gone awry after the merger with main. Test have been manually checked for consistency.

Commit:
Updated some remainding references to regulated_lists instead o control_lists in the default yaml inputs.

Commit:
Updated early exists for when no control list is detected to avoid overly unregulated and uncontrolled runs where nothing is controlled and everything passes, minor updates and tidying to the config file, and addition of fake control lists folder for functional tests to pass.

Commit:
Refactored yaml logic to allow for commec list to access yaml defaults, to allow users to pass -y to commec list, in lieu of -d (databases dir).

Commit:
Updated output csv for commec list to always have correct suffix, included Fungi as a valid control list category, and updated logic for setting preferred name to treat empty strings as None as intended.

Commit:
Updated the error messages and early exit for commec screen to handle gracefully the exits needed for when control lists are not found, and controlling for whether skip tx or not is used. Control list is deferred in the case of skip-tx to minimise verbosity of error handling during control list import failure.

Commit:
Fixed bugs with Taxids being treated as ints, not strings, lineage ingestion safety, and Toxin label for Category. Minor additional debug comments.

Commit:
Merge branch 'develop' into dynamic-annotations

Commit:
initial test implemnetations via Claude, some issues with ints() in tax_id imports still, all added tests require additional manual checking for sanity and purpose.

Commit:
Ensured imported nan values are text , not float nans which corrupted taxid text parsing. Fixed a domain vs domains grabbing bug from the annotations json output, removed the check for the existance of the no longer used biorisk controlled taxids csv path.

Commit:
Fixed an issue with domains not being passed into json hit descriptions correctly.

Commit:
Fixed some log messages and repeating taxid non regulated info.

Commit:
Fixed bug related to float reset on clear() command with control list module, as well as issue with child comparison not using the hash object.

Commit:
Removes int64 conversion in list data for ignored taxids

Commit:
Ommited erroneous debugging print message

Commit:
Updated some of the outputs for commec list to say something when a provided accession is not present, as well as ensured commec list can be run in output mode only for convenience, merged with annotations testing, some of which require additional verbose checking for usefullness. Deleted some more obvious offenders.

Commit:
Minor updates to the cli input for -l/-o/-a inputs and -d/-y interpretation error code for commec list tests

Commit:
Removed some unnecessary tests for ignored accessions.

Commit:
Initial conversions to new format of control list, utilising display name, and list_item etc. Potential optimised strategy for is_regulated vs the are_regulated batch call. Still requires extensive testing with commec screen.

Commit:
Addressed several broken tests due to changes in control list annotation and info changes.

Commit:
Further addressing all pytest errors, functional test included. Removed some tests that were simply too verboose.

Commit:
Updates the functional test json output, includes the species and genus infomration as expected.

Commit:
Refactors the check reg path file, adds some tests for the refactored elements.

Commit:
Merge with latest develop, ensured maintenance of yaml io changes in the yaml io refactor.

Commit:
Restored old database blast instructions in commandline.

Commit:
Overhaul to the control list reporting systems for check reg path. Match ranges are not singular, clustered sorting of hits based on control list outcomes, json output arrangement for taxonomy annotations overhaul. Still missing screen logic carrying, as well as html output correction.

Commit:
Updates for debugging and handling of the ignored taxa, as well as performance improvements to the ordering of operations during check reg path and taxid row explosion.

Commit:
Updates containers for control list category to support latest control lists.

Commit:
Updates to ensure the Dynamic annotations are working, final step is to ensure that those lists that re not under full compliance are correctly treated i.e. regional parsing.

Commit:
Updated to include initial regional differentation and to take into account list mode.

Commit:
Tidied some logic and outputs, some additional parsing requred for html to take into account more than the mixed outputs when taxonomy is a pass or warn for example. Otherwise working as intended.

Commit:
Removed orphaned code from check reg pathogens.

Commit:
Several updates to the testing environment and minor bug fixes to check reg path, namly changing the location of title collapsing to ensure that it occurs after clustering.

Commit:
Minor updates to some of the descriptions and entries for functional tests.

Commit:
Updates to include the NT coord conversion, as well as tweaks to the cluster coordinates represented on output, and after significant checking of the functional test output, and minor alterations, updated the functional expectations.

Commit:
working version of dyna anno using latest multiquery taxid retrieval and parsing. Several tests are now broken and need reworking. Coorected numerous bugs in the entire control list to hit to output pipeline, particular around child accession parsing and clustering correctness. Consider changing how clustering occurs, and need to remove the name and mapping of the hits array into a list rather than dict.

Commit:
Updated the Mixed flag results back to PASS, instead of warning. Which removes it from the rationale text for now. Updated a minor issue with the flag tests to pass after updating them introduced a typo.

Commit:
Initial changes to store hits as a list, instead of labelled dict, with
duplicated name field, which should look clearer. No longer require dict
as unique keys are not longer used as a method of hit deduplication.

Commit:
Updated all hits logic in the JSON to be a list, rather than mapped key with the hit name, the hit name has been changed to be the display name of the cluster where appropriate.

Commit:
Added the deduplication of low concern hits based on simple HitResult hashing.

Commit:
Updated functional test to not include duplicated low concern hit, as well as incremented the json version number to indicate the new hit storage as a list.

Commit:
Updated import, and tests, to not include notes column, as that output is deprecated for control lists, and tidied the output csv summary for control list cli to csv output.

Commit:
Removed match start and end coordinates from the MatchRange object, which is only used as the major HitResult coordinate system. As it was messily forcing the use when dealing witha Taxonomy hit, which by definition may have multiple match starts and ends. That information is now corrected put into the custom annotation fields for each element making up a Taxonomy hit. Biorisk and benign hits did not need to report this information anyway, as the user should not have access to the sequences to look up from those databases.

Commit:
Folded the data functions back into blast_tools. One day when we move away from blast we will no doubt make these independant of hteir tool, which will be as simple as passing in the headings for the dataframe as arguments, but they can live there for now or simplicity as they are only technically used for blast related processing.

Commit:
Updated how Non-coding regions work, they now are parsed separately, using the _X system, and the non-coding regions are correctly separated in the blast results interpretation to remove the possibility of a non coding hit transcending spacetime and covering across two very separate nc regions.

Commit:
Minor update to how the logging occurs for NT taxnomy.

Commit:
Fixed issue with Protein hits from other queries trimming the actual protein hits for that query, resulting in too many non-coding regions.

Commit:
Added the non-coding region calculation threshold to constants. It is currently set to 90.

Commit:
Merge with latest develop.

Commit:
in progress merge fixes.

Commit:
Added dummy control list data to be added to allow tests that require not hits to not fail on commec list importing. Dummy data is only added when needed.

Commit:
Tidied unused functions and erroneous comments.

Commit:
Tidying of check reg path, remove some unused functions and reordered fn calls.

Commit:
Added the hard-coded way to get rid of especially pesky accessions, starting with the measles coded HIV entry.

Commit:
Gutted all references to taxonomy database, vax taxids or synthetic taxids files, as they are now handled by control lists.

Commit:
Sort the lists in the output log such that the most important in terms of activation are at the top of the list.

Commit:
Merge branch 'develop' into dynamic-annotations

Commit:
Updated the use of the low concern annotations files to the csv one, which is likely the correct alternative as it features more annotations, it just needed some altering to get the description column to be treated as one large string. Which will occur in the latest low concern db.

Commit:
Add in the testing csv

Commit:
Ensured the new database arguments are actually applied.

Commit:
Removed double log message.

Commit:
Fixed the use of get top hits in benign synbio parts applying across queries.

Commit:
minor fix for non-existing low concern dna top hit getting.

Commit:
Fixed final issue, with more robust handling of problematic fasta names such as A_295 that occur in some test fastas, we now bypass issues with _X by trying queryname first on retrieval and doing token fallback when required only, rather than assuming.

Commit:
Reworded unused rationale text to be less confusing in the future.

Commit:
Updated commec list output csv to take into account internal tooling and therefore child relationships.

Commit:
Just alphabetized the output of control list -o
